### ISSUE TITLE
Introduce an additional assume operator `⩪` (dotsim) for varying dimensional models

### ISSUE THREAD

**Issue by yebai** [TuringLang/Turing.jl#708]
At the moment, all random variables in Turing models are declared through the tilde notation (`∼`). This is also true for variables that may dynamically appear and/or disappear during runtime (i.e. varying dimensional models). This creates many headaches for creating a consistent, intuitive scheme to re-name variables in `VarInfo`. 

Perhaps we should reconsider this issue and its solution. For example, if we introduce an additional assume operator `⩪`, and let

1. `∼` declares variables which always exist during runtime
2. `⩪` declares variables which may dynamically exist during runtime 

Then, for variables declared by (1), we always use the variable name (maybe plus some additional statically generated identifier) as their internal index in `VarInfo`. While, for variables declared by (2), we explicitly require the user to pass in a unique identifier through some API. 

This way, we maintain 99% per cent of the convenience of automatically generating variable names, but still, allow the user to write models with dynamic dimensionality and/or stochastic control flow.   

Related:

**Comment by yebai** [comment]
Ps. This would also enable the compiler to generate a type stable `VarInfo` since variables declared by `∼` are essentially static.

**Comment by willtebbutt** [comment]
My intuition is that we aught to be able to achieve this without introducing an extra assume operator, but I haven't thought this through sufficiently well to be able to provide a useful and informed opinion.

**Comment by yebai** [comment]
> My intuition is that we aught to be able to achieve this without introducing an extra assume operator, but I haven't thought this through sufficiently well to be able to provide a useful and informed opinion.

That's right. It's possible to use `~` everywhere with some additional arguments. For exposition reasons, let's assume we introduce `⩪`. 

> I guess I'm largely concerned that this will be confusing for the user. In particular, what happens if the user uses a regular sim when they should be using dotsim? 

There will be some checks for variables declared through regular sim. That is, these variables must exist in `VarInfo` and only exist once. Otherwise, inference will stop with an error. 

> How might this work with composable modelling? i.e. what if an entire model is dotsim-ed inside another model?

This can be addressed through compositional naming. For example, if `Foo` is a user-defined composite distribution, then all its internal variables will be prefixed with outer layer variable name. The following two examples show how this might work. 

Compositional modelling example:
```julia
@model Qux = begin x ~ Foo(); y ~ Bar();  end

@model Baz = begin
    # some stuff
    y ~ Qux() # Variable renamed to `y.x`, `y.y` in `VarInfo`
    # more stuff
end

@model Thud = begin
    # some stuff
    y ~ Baz() # Variable renamed to `y.y.x`, `y.y.y` in `VarInfo`
    # more stuff
end
```

More complicated case:
```julia
@model Foo = begin x ~ Foo()  end

@model Baz = begin
   # some stuff
   n ~ Poisson(5)
   for i = 1:n
      y ~ Foo() # Leads to runtime error if `z!=1`, cause there will be no or multiple `y.x`'s in `VarInfo`.
      z ⩪ Foo() # Compiler knows there might be varying copies of z.
  end
  # more stuff
end
```

I think we should only allow users to declare random variables by the `~`, `⩪` notations. That is,

```julia
# Forbid implicit declaration of random variable `x` in `Baz`
f() = x ~ Foo(); y ~ Bar();

@model Baz = begin
    # some stuff
    f() # Incorrect!
    # more stuff
end
```

This design might feel restrictive but makes it easier to reason about models, since it clearly distinguish between calling deterministic `functions`, and declaring random variables.

Any commments are appreciated!

**Comment by mohamed82008** [comment]
Having a way to differentiate between statically sized parameters that are always there and other parameters which are more dynamic is going to help with the performance. Currently, when handling parameters in `VarInfo`, we "find" the parameter in the `vals` vector by first generating a `VarName` from the symbol and runtime indices, and then using a dictionary `idcs` to find the index of this variable in the `vals` vector. For static parameters, we can make the underlying `vals` for this variable of the same shape as the parameter, thus enabling direct indexing into the array, avoiding the need for the `idcs` dictionary. This is possible after #660 which handles the `vals` for each symbol differently.

That said, I don't like the dotsim solution as it can create some confusion with the broadcast dot, and is IMO not very intuitive. I propose we have a static variable declaration in the model instead. Where the user does `@static x = Vector{Float64}(undef, 3)`.
